cirq/contrib/acquaintance/permutation.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Tuple, TypeVar
+
+from cirq import abc
+from cirq.ops import (
+        CompositeGate, Gate, QubitId, OP_TREE, SWAP,
+        flatten_op_tree, GateOperation, TextDiagrammable,
+        gate_features)
+
+LogicalIndex = TypeVar('LogicalIndex', int, QubitId)
+LogicalGates = Dict[Tuple[LogicalIndex, ...], Gate]
+LogicalMappingKey = TypeVar('LogicalMappingKey', bound=QubitId)
+LogicalMapping = Dict[LogicalMappingKey, LogicalIndex]
+
+class PermutationGate(Gate, TextDiagrammable, metaclass=abc.ABCMeta):
+    """Permutation gate."""
+
+    def __init__(self, swap_gate: Gate=SWAP) -> None:
+        self.swap_gate = swap_gate
+
+
+    @abc.abstractmethod
+    def permutation(self, qubit_count: int) -> Dict[int, int]:
+        """permutation = {i: s[i]} indicates that the i-th qubit is mapped to
```

Comment:
It's strange how the swap gate is passed as an argument to the constructor, but the permutation is passed via inheritance. Could they both be passed as arguments?

Implementing `Operation` instead of `Gate` might help with that.

Reply:
Except for LinearPermutationGate, the permutation is usually fixed by the subclass, so I don't see why we would want to pass it as an argument to the initializer.

Reply:
To be honest, I still don't understand the move from the `Operation` = `Gate` + `Qubits` model. It seems like now we just need to create an additional `Operation` subclass to indicate the type of gate. What's the benefit?

Reply:
The benefit comes when the gate has to talk in detail about the qubits. For example, a QubitOp associates various qubits with X, Y, or Z axes. If you implement a gate, which does not know its qubits, you need to talk about this association in an indirect way (e.g. via integer indices), or you need to require that the user pass in all of the qubits associated with axes when creating the operation, or various other hacky things to get around the fact that gates can't talk about quits directly. This creates many opportunities for mistakes.

With an Operation, you always know what your qubits are so it's much more natural.